lib/pbench/test/unit/agent/test_tool_meister.py

```diff
@@ -139,60 +135,47 @@ def f(directory: pathlib.Path, tar_file: pathlib.Path):
-        return io.StringIO()
+        def fake_unlink(*args):
+            assert args[0] == pathlib.Path(f"{self.directory}.tar.xz")
+            self.functions_called.append("fake_unlink")
 
-    def fake_rmtree(self, directory: pathlib.Path):
-        assert directory == tmp_dir
-        self.function_called.append("fake_rmtree")
-        pass
+        def fake_open(*args):
+            assert args[0] == pathlib.Path(f"{self.directory}.tar.xz")
+            self.functions_called.append("fake_open")
+            return io.StringIO()
 
-    def fake_md5(self, tar_file: pathlib.Path):
-        assert tar_file == pathlib.Path(f"{self.directory}.tar.xz")
-        self.function_called.append("fake_md5")
-        return 10, "random_md5"
+        def fake_rmtree(directory: pathlib.Path):
+            assert directory == tmp_dir
+            self.functions_called.append("fake_rmtree")
 
-    @responses.activate
-    def test_tar_create_success(self, monkeypatch):
-        """This test should pass the tar creation in send directory"""
+        def fake_md5(tar_file: pathlib.Path):
+            assert tar_file == pathlib.Path(f"{self.directory}.tar.xz")
+            self.functions_called.append("fake_md5")
+            return 10, "random_md5"
 
-        monkeypatch.setattr(shutil, "rmtree", self.fake_rmtree)
-        monkeypatch.setattr("pbench.agent.tool_meister.md5sum", self.fake_md5)
-        monkeypatch.setattr(pathlib.Path, "unlink", self.fake_unlink)
-        monkeypatch.setattr(pathlib.Path, "open", self.fake_open)
-
-        tm = ToolMeister(
-            pbench_install_dir=None,
-            tmp_dir=None,
-            tar_path="tar_path",
-            sysinfo_dump=None,
-            params=TestCreateTar.tm_params,
-            redis_server=None,
-            logger=logging.getLogger(),
-        )
+        monkeypatch.setattr(shutil, "rmtree", fake_rmtree)
+        monkeypatch.setattr("pbench.agent.tool_meister.md5sum", fake_md5)
+        monkeypatch.setattr(pathlib.Path, "unlink", fake_unlink)
+        monkeypatch.setattr(pathlib.Path, "open", fake_open)
 
-        monkeypatch.setattr(tm, "_create_tar", self.fake_create_tar(0, b""))
+        monkeypatch.setattr(tool_meister, "_create_tar", self.fake_create_tar(0, b""))
 
         url = (
-            f"http://{TestCreateTar.tm_params['tds_hostname']}:{TestCreateTar.tm_params['tds_port']}/uri"
-            f"/ctx/{TestCreateTar.tm_params['hostname']}"
+            f"http://{tm_params['tds_hostname']}:{tm_params['tds_port']}/uri"
+            f"/ctx/{tm_params['hostname']}"
         )
-        self.add_http_mock_response(url)
+        responses.add(responses.PUT, url, status=HTTPStatus.OK, body="succeeded")
 
-        failures = tm._send_directory(self.directory, "uri", "ctx")
-        assert self.function_called == [
+        failures = tool_meister._send_directory(self.directory, "uri", "ctx")
+        functions_called, self.functions_called = self.functions_called, []
```

Comment:
Use of `nonlocal` makes sense.

Reply:
I don't know that `nonlocal` will "reach".  (I know that Python has "interesting" scoping rules, but I'm not sure that the invocation of a static class method can reference a local variable in its caller's scope with a simple `nonlocal` definition, because, lexically, the caller is not a "containing" scope...but, I don't know what the dynamic scoping rules are.)

On the flip side, though, I don't think it's a problem for the inner function of `fake_create_tar()` to access the outer function's parameters, and Python provides "pass by name" semantics for parameters, so the mock could have access to the local list without the use of `nonlocal` if we just pass it in and let it be part of the closure.

Reply:
I spoke with @npalaska, this works just fine with a local variable, `nonlocal` is not needed.